File: 每日一题/4.30快乐数.py
'''
法一：使用hashset检查循环
'''
class Solution:
    def isHappy(self, n: int) -> bool:
        
        def getNext(n):
            totalSum = 0
            while n > 0:
                n, remains = divmod(n, 10)
                totalSum += remains ** 2
            return totalSum
        
        seen = set()
        while n != 1 and n not in seen:
            seen.add(n)
            n = getNext(n)

        return n == 1


'''
法二：模拟快慢指针
'''
class Solution:
    def isHappy(self, n: int) -> bool:

        def getNext(n):
            totalSum = 0
            while n > 0:
                n, remains = divmod(n, 10)
                totalSum += remains ** 2
            return totalSum

        slow = n
        fast = getNext(n)
        while fast != 1 and slow != fast:
            slow = getNext(slow)
            fast = getNext(getNext(fast))
        
        return fast == 1

File: 每日一题/4.30快乐数.cpp
/*
法一：使用hashset检测指针
*/
class Solution {
    int computeNumber(int n) {
        int res = 0;
        while (n) {
            res += pow(n % 10, 2.0);
            n /= 10;
        }
        return res;
    }

public:
    bool isHappy(int n) {
        set<int> numbers;
        int temp = n;
        while (1) {
            temp = this->computeNumber(temp);
            if (temp == 1) return true;
            if (temp == 0 || numbers.find(temp) != numbers.end()) {
                break;
            }
            numbers.insert(temp);
        }
        return false;
    }
};

/*
法二：快慢指针法
*/
class Solution {
    int getNext(int n) { // 得到下一个数
        int totalSum = 0;
        while (n > 0) {
            int d = n % 10;
            n = n / 10;
            totalSum += d * d;
        }
        return totalSum;
    }

public:
    bool isHappy(int n) {
        int slow = n;
        int fast = getNext(n);
        while (fast != 1 && slow != fast) {
            slow = getNext(slow);
            fast = getNext(getNext(fast));
        }
        return fast == 1;
    }
};